File: fetch_skills.js
/**
 * fetch_skills.js
 * 
 * 執行方式：node fetch_skills.js
 * 
 * 功能：直接連線 QuestLog API 抓取全量技能資料，
 *      生成靜態資料檔案 js/skill_data_static.js。
 * 
 * 📌 以後遊戲新增技能需要更新時，只要執行：
 *    node fetch_skills.js
 *    然後 commit js/skill_data_static.js 即可。
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

// ── 讀取 skill-names.js 取得全量技能 ID 清單 ──
const skillNamesContent = fs.readFileSync(path.join(__dirname, 'js', 'skill-names.js'), 'utf8');
const skillIdMatch = skillNamesContent.match(/"(\d{8})":\s*"([^"]+)"/g);
const skillsToFetch = {};

if (skillIdMatch) {
    skillIdMatch.forEach(m => {
        const parts = m.match(/"(\d{8})":\s*"([^"]+)"/);
        if (parts) skillsToFetch[parts[1]] = parts[2];
    });
}

const results = {};

// ── 直接打 QuestLog API（Node.js 不受 CORS 限制）──
function fetchSkill(id) {
    return new Promise((resolve) => {
        const input = encodeURIComponent(JSON.stringify({ id: id.toString(), language: 'zh' }));
        const url = `https://questlog.gg/aion-2/api/trpc/database.getSkill?input=${input}`;

        const options = {
            headers: {
                'User-Agent': 'Mozilla/5.0',
                'Accept': 'application/json'
            },
            timeout: 10000
        };

        https.get(url, options, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => {
                try {
                    const json = JSON.parse(data);
                    const rawData = json.result?.data?.json || json.result?.data || json.data;
                    resolve(rawData || null);
                } catch (e) { resolve(null); }
            });
        }).on('error', () => resolve(null))
            .on('timeout', function () { this.destroy(); resolve(null); });
    });
}

async function run() {
    const ids = Object.keys(skillsToFetch);
    console.log(`🚀 開始抓取全量技能 (共 ${ids.length} 個)...\n`);

    let ok = 0, fail = 0;

    for (let i = 0; i < ids.length; i++) {
        const id = ids[i];
        process.stdout.write(`[${i + 1}/${ids.length}] ${skillsToFetch[id]} (${id})... `);

        let data = null;
        // 失敗時最多重試 2 次
        for (let retry = 0; retry < 3; retry++) {
            data = await fetchSkill(id);
            if (data) break;
            if (retry < 2) await new Promise(r => setTimeout(r, 500));
        }

        if (data) {
            results[id] = {
                name: data.name,
                icon: data.icon,
                descriptionData: data.descriptionData,
                levels: data.levels
            };
            console.log('✅');
            ok++;
        } else {
            console.log('❌ 無資料（可能此技能不在 QuestLog 資料庫中）');
            fail++;
        }

        await new Promise(r => setTimeout(r, 120)); // 避免頻率過高
    }

    console.log(`\n🎉 完成！成功: ${ok} 個，失敗: ${fail} 個`);
    const output = `window.SKILL_DATA_STATIC = ${JSON.stringify(results, null, 2)};\n`;
    fs.writeFileSync(path.join(__dirname, 'js', 'skill_data_static.js'), output, 'utf8');
    console.log('✅ 已儲存至 js/skill_data_static.js');
    console.log('📌 請 commit 這個檔案後 push 到 GitHub Pages 即可生效。');
}

run();

File: fetch_skills.py

import json
import urllib.request
import urllib.parse
import time
import os

# 1. 定義要抓取的技能名稱 (從 passive_skills.js 翻譯)
target_skill_names = [
    # 劍星
    "生存姿態", "保護盔甲", "掌握弱點", "攻擊準備", "衝擊擊中",
    # 守護星
    "體力強化", "庇護盾牌", "斷罪加護", "銅牆鐵壁", "侮辱咆哮", 
    # 殺星
    "第六感最大化", "瞄準破綻", "背後強擊", "強襲姿態",
    # 弓星
    "獵人決心", "警戒之眼", "集中之眼", "抵抗決心", "回生契約",
    # 魔道星
    "大地長袍", "火花長袍", "抵抗恩惠", "強化恩惠",
    # 精靈星
    "精靈打擊", "精靈保護", "侵蝕", "精神集中",
    # 治癒星
    "溫慢加護", "主神加護", "主神恩寵", "治癒力強化", "不死帳幕", "生存意志",
    # 護法星
    "生命祝福", "十字防禦", "鼓吹咒語", "激怒咒語", "大地約定", "風之約定"
]

# 2. 從 skill-names.js 讀取 ID 映射
# 這裡手動建立一些關鍵映射，或掃描檔案 (為了精確性，我直接列出剛才看到的)
# 實際上我應該從檔案讀取
skill_names_db = {
    "11710000": "生存姿態", "11720000": "保護盔甲", "11740000": "掌握弱點", "11750000": "攻擊準備", "11760000": "衝擊擊中",
    "12710000": "體力強化", "12720000": "庇護盾牌", "12730000": "斷罪加護", "12740000": "銅牆鐵壁", "12760000": "衝擊擊中", "12770000": "侮辱咆哮", "12790000": "生存意志",
    "13710000": "第六感最大化", "13720000": "瞄準破綻", "13740000": "背後強擊", "13750000": "強襲姿態", "13760000": "衝擊擊中",
    "14710000": "警戒之眼", "14740000": "集中之眼", "14750000": "獵人決心", "14760000": "抵抗決心", "14790000": "回生契約",
    "15720000": "大地長袍", "15740000": "火花長袍", "15770000": "抵抗恩惠", "15780000": "強化恩惠", "15790000": "回生契約",
    "16710000": "精靈打擊", "16720000": "精靈保護", "16740000": "侵蝕", "16760000": "精神集中", "16790000": "回生契約",
    "17710000": "溫暖加護", "17720000": "主神加護", "17730000": "主神恩寵", "17740000": "治癒力強化", "17750000": "不死帳幕", "17790000": "生存意志",
    "18710000": "生命祝福", "18720000": "十字防禦", "18740000": "鼓吹咒語", "18750000": "攻擊準備", "18760000": "衝擊擊中", "18770000": "激怒咒語", "18780000": "大地約定", "18800000": "風之約定"
}

def fetch_skill(skill_id):
    url = f"https://questlog.gg/aion-2/api/trpc/database.getSkill?input=%7B%22id%22%3A%22{skill_id}%22%2C%22language%22%3A%22zh%22%7D"
    try:
        with urllib.request.urlopen(url) as response:
            data = json.loads(response.read().decode())
            return data.get("result", {}).get("data", {}).get("json")
    except Exception as e:
        print(f"Error fetching {skill_id}: {e}")
        return None

results = {}
for skill_id, name in skill_names_db.items():
    print(f"Fetching {name} ({skill_id})...")
    data = fetch_skill(skill_id)
    if data:
        results[skill_id] = data
    time.sleep(0.3) # 避免太快

output_path = r"d:\c150075\Desktop\aiongame\js\skill_data_static.js"
with open(output_path, "w", encoding="utf-8") as f:
    f.write("window.SKILL_DATA_STATIC = ")
    json.dump(results, f, ensure_ascii=False, indent=2)
    f.write(";")

print(f"Done! Saved to {output_path}")
